xarray/core/dataset.py

```diff
@@ -2980,20 +2980,30 @@ def isel(
         coord_names = self._coord_names.copy()
 
         indexes, index_variables = isel_indexes(self.xindexes, indexers)
+        all_keys = set(indexers.keys())
 
         for name, var in self._variables.items():
             # preserve variable order
             if name in index_variables:
                 var = index_variables[name]
-            else:
-                var_indexers = {k: v for k, v in indexers.items() if k in var.dims}
-                if var_indexers:
+                dims.update(zip(var.dims, var.shape))
+            # Fastpath, skip all of this for variables with no dimensions
+            # Keep the result cached for future dictionary update
+            elif var_dims := var.dims:
```

Comment:
oh even faster!!!

6.1k -> 6.6k!!

Reply:
no wait, i spoke too soon. i had a typo. oddly, it is slower...

Reply:
```patch
diff --git a/xarray/core/dataset.py b/xarray/core/dataset.py
index ec756176..4e8c31e5 100644
--- a/xarray/core/dataset.py
+++ b/xarray/core/dataset.py
@@ -2987,22 +2987,20 @@ class Dataset(
             if name in index_variables:
                 var = index_variables[name]
                 dims.update(zip(var.dims, var.shape))
-            # Fastpath, skip all of this for variables with no dimensions
-            # Keep the result cached for future dictionary update
-            elif var_dims := var.dims:
+            elif var.ndim == 0:
+                continue
+            else:
                 # Large datasets with alot of metadata may have many scalars
                 # without any relevant dimensions for slicing.
                 # Pick those out quickly and avoid paying the cost below
                 # of resolving the var_indexers variables
-                if var_indexer_keys := all_keys.intersection(var_dims):
+                if var_indexer_keys := all_keys.intersection(var.dims):
                     var_indexers = {k: indexers[k] for k in var_indexer_keys}
                     var = var.isel(var_indexers)
                     if drop and var.ndim == 0 and name in coord_names:
                         coord_names.remove(name)
                         continue
-                    # Update our reference to `var_dims` after the call to isel
-                    var_dims = var.dims
-                dims.update(zip(var_dims, var.shape))
+                dims.update(zip(var.dims, var.shape))
             variables[name] = var

         return self._construct_direct(
```

was slower.... this is somewhat unexpected. ndim should be "instant".

Reply:
let me add a benchmark tonight to "show" that this is the better way explicitely, otherwise it will be too easy to undo.

Reply:
The reason for this slowdown is that `def dims` -> returns `self._dims` is actually a property of the class.

`ndim` is typically provided by the ndarraymixin, which ends up resolving 3 layers down to `len(self.shape)`, 

yo ucould use `elif var.shape` (which improves a bit compared to the var.ndim) test, but the `elif var_dims := var.dims` is still the fastest.

Reply:
> var_dims := var.dims is still the fastest.

Isn't this only true when you're repeatedly iterating over scalars. If we happened to iterate alternating between 0D and 1D vars we'd lose the optimization?

I'd prefer `elif var.shape` with a comment explaining why not to use `.ndim` since it's more readable and much easier to understand at first glance.

Reply:
My conclusion is that:

* `len(tuple)` seems to be pretty fast.
* But the `.shape` attribute is only resolved after 4-5 different python indirections going down to a `LazilyIndexedArray`, `MemoryCachedArray`, `H5BackedArray` (sorry, i'm not getting the class names right), but ultimately it isn't "readily available and needs to be resolved.

My little heuristic test is that with my dataset (93 variables long)

```python
In [16]: %%timeit
    ...: for v in dataset._variables.values():
    ...:     v.ndim
    ...:
119 µs ± 1.17 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [17]: %%timeit
    ...: for v in dataset._variables.values():
    ...:     v.shape
    ...:
105 µs ± 1.09 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [18]: %%timeit
    ...: for v in dataset._variables.values():
    ...:     v.dims
    ...:
7.66 µs ± 38.4 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

In [19]: %%timeit
    ...: for v in dataset._variables.values():
    ...:     v._dims
    ...:
3.1 µs ± 22 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

In [20]: len(dataset._variables)
93
```

I mean, micro-optimizations are sometimes dumb. So that is why i've been breaking them out into distinct ideas when I find them, but together they can add up, especially when taken together.

So in other words, my hypothesis is that the the use of `_dims` is really helpful because it avoids many indirections in `shape` since dims is a "cached" version of the shape (where every number is replaced with a string).

Reply:
`len(v.dims)` or `len(v._dims)` sounds OK to me. They're both readily understandable

Reply:
Just so I better understand the xarray style,

The truthyness of tuples is not obvious enough while Len(tuple) is more obviously associated with a true/false statement

Would a comment be ok if Len(tuple) hurts performance?

Reply:
It not about style, but about readability and understandability.

I've read this snippet about 6 times now, but I still have to look at it closely to see what it does. The perf improvement is also sensitive to order of iteration over variables (what if you alternated between 0D and 1D variable as you iterated through?)

This is why I'd prefer an explicit check for scalar variable. It's easy to see and reason about the special-case.

Reply:
> what if you alternated between 0D and 1D variable as you iterated through?

You know, this is something I've thought about alot.

I'm generally not too happy with this optimization.

> This is why I'd prefer an explicit check for scalar variable. It's easy to see and reason about the special-case.

Ok understood. The challenge is that this PR doesn't do much on my benchmarks without https://github.com/pydata/xarray/pull/9002 and my current theory is that we are limited by calls to python methods, so I feel like even `len(tuple)` will slow things down.

I'll try again, but if its OK, i'm going to rebase onto #9002 until a resolution is found for those optimizations.